File: iosApp/iosApp/Permission/PermissionScreen.swift
import SwiftUI
import CoreLocation

struct PermissionScreen: View {

    let isDenied: Bool
    let onRequestPermission: () -> Void
    let onOpenSettings: () -> Void

    @State private var isAwaitingPermission = false
    @StateObject private var locationRequester = LocationPermissionRequester()

    var body: some View {
        ZStack {
            content

            if isAwaitingPermission {
                Color.appBackground.ignoresSafeArea()
                VStack(spacing: 16) {
                    ProgressView().scaleEffect(1.4).tint(.appPrimary)
                    Text("Verificando permissão...")
                        .font(.appBodySm)
                        .foregroundColor(.appTextMuted)
                }
            }
        }
        .onAppear {
            locationRequester.onAuthorizationChanged = {
                isAwaitingPermission = false
                onRequestPermission()
            }
        }
    }

    private var content: some View {
        ScrollView {
            VStack(spacing: 0) {
                Spacer().frame(height: 60)

                Text("📍")
                    .font(.system(size: 64))

                Spacer().frame(height: 24)

                Text(isDenied ? "Permissão negada" : "Permissão de localização")
                    .font(.appH3)
                    .foregroundColor(.appTextDark)
                    .multilineTextAlignment(.center)

                Spacer().frame(height: 12)

                Text(
                    isDenied
                        ? "Você negou o acesso à localização. Para usar o TEMBT você precisa habilitá-la nas configurações do iPhone."
                        : "O TEMBT precisa da sua localização para funcionar. Veja abaixo o que é solicitado e por quê."
                )
                .font(.appBody)
                .foregroundColor(.appTextMuted)
                .multilineTextAlignment(.center)
                .padding(.horizontal, 32)

                Spacer().frame(height: 28)

                if isDenied {
                    SettingsPathCard()
                        .padding(.horizontal, 24)
                } else {
                    PermissionExplanationCard()
                        .padding(.horizontal, 24)
                }

                Spacer().frame(height: 32)

                Button(isDenied ? "Abrir configurações" : "Permitir localização") {
                    if isDenied {
                        if let url = URL(string: UIApplication.openSettingsURLString) {
                            UIApplication.shared.open(url)
                        }
                        onOpenSettings()
                    } else {
                        isAwaitingPermission = true
                        locationRequester.requestPermission()
                    }
                }
                .buttonStyle(.borderedProminent)
                .tint(.appPrimary)
                .font(.appTitle)
                .padding(.horizontal, 32)
                .frame(maxWidth: .infinity)

                Spacer().frame(height: 60)
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(Color.appBackground)
    }
}

// MARK: - Explanation card (not denied)

private struct PermissionExplanationCard: View {
    var body: some View {
        VStack(alignment: .leading, spacing: 16) {
            Text("PERMISSÃO NECESSÁRIA")
                .font(.system(size: 11, weight: .bold))
                .foregroundColor(.appPrimary)
                .tracking(0.8)

            Text("Localização — \"Ao usar o app\"")
                .font(.system(size: 15, weight: .semibold))
                .foregroundColor(.appTextDark)

            VStack(alignment: .leading, spacing: 12) {
                PermissionReasonRow(
                    icon: "🗺️",
                    text: "Mostrar você e outros jogadores no mapa da quadra em tempo real."
                )
                PermissionReasonRow(
                    icon: "📅",
                    text: "Detectar automaticamente se você está na quadra nos fins de semana e feriados, e atualizar sua presença em segundo plano."
                )
            }
        }
        .padding(20)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(
            RoundedRectangle(cornerRadius: 12)
                .fill(Color(.secondarySystemBackground))
        )
    }
}

// MARK: - Settings path card (denied)

private struct SettingsPathCard: View {
    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            Text("COMO HABILITAR")
                .font(.system(size: 11, weight: .bold))
                .foregroundColor(.appPrimary)
                .tracking(0.8)

            Text("Ajustes  →  TEMBT  →  Localização  →  \"Ao usar o app\"")
                .font(.system(size: 14))
                .foregroundColor(.appTextMuted)
        }
        .padding(20)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(
            RoundedRectangle(cornerRadius: 12)
                .fill(Color(.secondarySystemBackground))
        )
    }
}

// MARK: - Reason row

private struct PermissionReasonRow: View {
    let icon: String
    let text: String

    var body: some View {
        HStack(alignment: .top, spacing: 12) {
            Text(icon)
                .font(.system(size: 18))
                .frame(width: 24)
            Text(text)
                .font(.appBodySm)
                .foregroundColor(.appTextMuted)
                .fixedSize(horizontal: false, vertical: true)
        }
    }
}

// MARK: - Permission requester

@MainActor
final class LocationPermissionRequester: NSObject, ObservableObject, CLLocationManagerDelegate {

    var onAuthorizationChanged: (() -> Void)?

    private let manager = CLLocationManager()

    override init() {
        super.init()
        manager.delegate = self
    }

    func requestPermission() {
        manager.requestWhenInUseAuthorization()
    }

    nonisolated func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
        Task { @MainActor in
            let status = manager.authorizationStatus
            if status != .notDetermined {
                onAuthorizationChanged?()
            }
        }
    }

    deinit {
        manager.delegate = nil
    }
}

File: androidApp/src/main/kotlin/com/tembt/android/ui/permission/PermissionScreen.kt
package com.tembt.android.ui.permission

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import com.tembt.domain.model.LocationPermissionStatus

@Composable
fun PermissionScreen(
    permissionStatus: LocationPermissionStatus,
    onRequestPermission: () -> Unit,
    onOpenSettings: () -> Unit
) {
    val isDenied = permissionStatus == LocationPermissionStatus.DENIED

    Surface(
        modifier = Modifier.fillMaxSize(),
        color = MaterialTheme.colorScheme.background
    ) {
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(32.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.Center
        ) {
            Text(
                text = "📍",
                style = MaterialTheme.typography.displayLarge
            )

            Spacer(modifier = Modifier.height(24.dp))

            Text(
                text = if (isDenied) "Permissão negada" else "Permissão de localização",
                style = MaterialTheme.typography.headlineMedium,
                textAlign = TextAlign.Center
            )

            Spacer(modifier = Modifier.height(12.dp))

            Text(
                text = if (isDenied)
                    "Você negou o acesso à localização. Para usar o TEMBT você precisa habilitá-la nas configurações do dispositivo."
                else
                    "O TEMBT precisa da sua localização para funcionar. Veja abaixo o que é solicitado e por quê.",
                style = MaterialTheme.typography.bodyLarge,
                textAlign = TextAlign.Center,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )

            Spacer(modifier = Modifier.height(28.dp))

            if (isDenied) {
                SettingsPathCard()
            } else {
                PermissionExplanationCard()
            }

            Spacer(modifier = Modifier.height(32.dp))

            Button(
                onClick = if (isDenied) onOpenSettings else onRequestPermission,
                modifier = Modifier.fillMaxWidth()
            ) {
                Text(text = if (isDenied) "Abrir configurações" else "Permitir localização")
            }
        }
    }
}

@Composable
private fun PermissionExplanationCard() {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(12.dp))
            .background(MaterialTheme.colorScheme.surfaceVariant)
            .padding(20.dp),
        verticalArrangement = Arrangement.spacedBy(16.dp)
    ) {
        Text(
            text = "PERMISSÃO NECESSÁRIA",
            style = MaterialTheme.typography.labelSmall,
            color = MaterialTheme.colorScheme.primary,
            fontWeight = FontWeight.Bold
        )

        Text(
            text = "Localização — \"Ao usar o app\"",
            style = MaterialTheme.typography.titleSmall,
            fontWeight = FontWeight.SemiBold
        )

        Column(verticalArrangement = Arrangement.spacedBy(10.dp)) {
            PermissionReasonRow(
                icon = "🗺️",
                text = "Mostrar você e outros jogadores no mapa da quadra em tempo real."
            )
            PermissionReasonRow(
                icon = "📅",
                text = "Detectar automaticamente se você está na quadra nos fins de semana e feriados, e atualizar sua presença em segundo plano."
            )
        }
    }
}

@Composable
private fun SettingsPathCard() {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(12.dp))
            .background(MaterialTheme.colorScheme.surfaceVariant)
            .padding(20.dp),
        verticalArrangement = Arrangement.spacedBy(12.dp)
    ) {
        Text(
            text = "COMO HABILITAR",
            style = MaterialTheme.typography.labelSmall,
            color = MaterialTheme.colorScheme.primary,
            fontWeight = FontWeight.Bold
        )

        Text(
            text = "Configurações  →  Apps  →  TEMBT  →  Permissões  →  Localização  →  \"Ao usar o app\"",
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant
        )
    }
}

@Composable
private fun PermissionReasonRow(icon: String, text: String) {
    Row(
        horizontalArrangement = Arrangement.spacedBy(12.dp),
        verticalAlignment = Alignment.Top
    ) {
        Text(text = icon, modifier = Modifier.size(20.dp))
        Text(
            text = text,
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant
        )
    }
}
